### DISCUSSION TITLE
jsPsych.endExperiment in JATOS

### DISCUSSION THREAD

**Discussion by kittymimo** [jspsych/jsPsych#1411] [Q&A] [answered]
Hi!

I have an experiment in which I use `jsPsych.endExperiment` with a different message for different trials. Here is an example:
```
	var q5 = {
    		type: "html-button-response",
    		stimulus:
    			"<p><b>Votre enfant habite-t-il(-elle) au Québec?</b></p>",
    		choices: ["Oui", "Non"],
    		on_finish: function(data){
			if(data.button_pressed == 1){
				jsPsych.endExperiment(
					"Pour être éligible à l’étude, votre enfant doit habiter au Québec.<br />"+
					"Votre enfant ne peut malheureusement pas participer à cette étude.<br />"+
					"Merci pour votre temps!"
				);
			}
			if(data.button_pressed == 0){
        			count++;
        			var progress = count/n_trials
            			jsPsych.setProgressBar(progress);
        		} 
		}
	};
	
	var q6 = {
    		type: "html-button-response",
    		stimulus:
    			"<p><b>La langue maternelle de votre enfant (la première langue qu’il <br />"+
    			"ou elle a apprise et comprend toujours) est-elle le français?</b></p>",
    		choices: ["Oui", "Non"],
    		on_finish: function(data){
			if(data.button_pressed == 1){
				jsPsych.endExperiment(
					"Pour être éligible à l’étude, votre enfant doit avoir le français comme langue maternelle.<br />"+
					"Votre enfant ne peut malheureusement pas participer à cette étude.<br />"+
					"Merci pour votre temps!"
				);
			}
			if(data.button_pressed == 0){
        			count++;
        			var progress = count/n_trials
            			jsPsych.setProgressBar(progress);
        		} 
		}
	};
```
This works perfectly when I run the experiment on my computer, but in JATOS, the message disappears immediately. This is my code:
```
	jatos.onLoad(function() {
		jsPsych.init({
			timeline: timeline,
        		preload_audio: audio,
        		preload_images: images,
 			use_webaudio: true,
 			show_progress_bar: true,
    			auto_update_progress_bar: false,
			message_progress_bar: "",
			on_finish: function() {
           			var resultats = jsPsych.data.get().csv();
           			jatos.endStudy(resultats);
       			}
		});
	});
```
Do you know what I could do to make this work in JATOS?

Thanks!

Catherine

**Comment by jodeleeuw** [comment]
Hi Catherine,

I don't have much experience working with JATOS, but I suspect that the `jatos.endStudy()` command is being called because it is in the `on_finish` event for the experiment. You could try moving this command into a plugin at the end of the timeline instead.

```js
var end_experiment = {
	type: 'call-function',
	func: function(){
		var resultats = jsPsych.data.get().csv();
           	jatos.endStudy(resultats);
	}
}
```

**Reply by kittymimo** [reply]
I ended up doing something different (conditional trials) but thanks for your help!

**Comment by becky-gilbert** [comment, marked as answer]
Hi Catherine,
The `jatos.endStudy` function will automatically redirect to their end study page, which means that any jsPsych content that you want to remain on the page will no longer be visible once that function is called. If you want to end the study in JATOS _without_ redirecting to their end page, you can use the [`jatos.endStudyAjax` function](http://www.jatos.org/jatos.js-Reference.html). Note that if you want to send the results back to JATOS before ending the study, then you should do something like example 5 in the JATOS endStudyAjax docs:
```javascript
var resultats = jsPsych.data.get().csv();
jatos.submitResultData(resultats)
  .then(jatos.endStudyAjax)
```
I hope this helps! And if you have any JATOS-related follow-up questions, feel free to post on the [JATOS google group](https://groups.google.com/g/jatos) 😃

**Reply by kittymimo** [reply]
Thanks! I ended up using conditional trials, but I'll try this function in the future. And thanks for letting me know about the JATOS Google group!
